File: src/test/java/org.dcache.xrootd4j/CmsMappingHandlerTest.java
package org.dcache.xrootd4j;

import org.junit.*;

public class CmsMappingHandlerTest {

    CmsMappingHandler handler;

    @Before
    public void setup() {

        handler = new CmsMappingHandler(CmsSettingsParser.parse(
                "<storage-mapping>\n" +
                        "\n" +
                        "  <pfn-to-lfn protocol=\"direct\"\n" +
                        "    path-match=\"/pnfs/desy.de/cms/analysis/dcms/unmerged/(.*)\"\n" +
                        "    result=\"/store/unmerged/DCMS/$1\"/>\n" +
                        "  <pfn-to-lfn protocol=\"direct\"\n" +
                        "    path-match=\"/pnfs/desy.de/cms/analysis/dcms/(.*)\"\n" +
                        "    result=\"/store/DCMS/$1\"/>\n" +
                        "  <pfn-to-lfn protocol=\"direct\"\n" +
                        "    path-match=\"/pnfs/desy.de/cms/tier2/loadtest/(.*)\"\n" +
                        "    result=\"/store/PhEDEx_LoadTest07/$1\"/>\n" +
                        "  <pfn-to-lfn protocol=\"direct\"\n" +
                        "    path-match=\"/pnfs/desy.de/cms/tier2/unmerged/(.*)\"\n" +
                        "    result=\"/store/unmerged/$1\"/>\n" +
                        "  <pfn-to-lfn protocol=\"direct\"\n" +
                        "    path-match=\"/pnfs/desy.de/cms/tier2/temp/(.*)\"\n" +
                        "    result=\"/store/temp/$1\"/>\n" +
                        "  <pfn-to-lfn protocol=\"remote-xrootd\"\n" +
                        "    path-match=\"/(.*)\"\n" +
                        "    result=\"/$1\"/>\n" +
                        "  <pfn-to-lfn protocol=\"direct\"\n" +
                        "    path-match=\"/pnfs/desy.de/cms/tier2/(.*)\"\n" +
                        "    result=\"/$1\"/>\n" +
                        "  <pfn-to-lfn protocol=\"srm\" chain=\"direct\"\n" +
                        "    path-match=\".*\\?SFN=(.*)\"\n" +
                        "    result=\"$1\"/>\n" +
                        "  <pfn-to-lfn protocol=\"srmv2\" chain=\"direct\"\n" +
                        "    path-match=\".*\\?SFN=(.*)\"\n" +
                        "    result=\"$1\"/>\n" +
                "</storage-mapping>"));
    }

    @Test
    public void shouldLeaveUnmatchedProtocolUnchanged() {
        String result =handler.authorize(null, null, null,
                "ftp://store/blubb",
                null, 0, null);

        Assert.assertEquals("ftp://store/blubb", result);
    }

    @Test
    public void shouldLeaveUnmatchedPathUnchanged() {
        String result =handler.authorize(null, null, null,
                "srm://unmatched/blubb",
                null, 0, null);

        Assert.assertEquals("srm://unmatched/blubb", result);
    }

    @Test
    public void shouldMapMatchingPath() {
        String result = handler.authorize(null, null, null,
                "remote-xrootd:///pnfs/fs/usr/cms/WAX/11/store/wurstbrot",
                null, 0, null);

        Assert.assertEquals("/pnfs/fs/usr/cms/WAX/11/store/wurstbrot", result);
    }

    @Test
    public void shouldMapWithChainedRule() {
        String result = handler.authorize(null, null, null,
                "srm://schwupp/ti/du?SFN=/pnfs/desy.de/cms/tier2/temp/wurstbrot",
                null, 0, null);

        Assert.assertEquals("/store/temp/wurstbrot", result);
    }
}

File: src/test/scala/org/dcache/xrootd4j/CmsMappingHandlerTest.scala
/**
 *  This file is part of xrootd4j-cms-plugin.
 *
 *  xrootd4j-cms-plugin is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  xrootd4j-cms-plugin is distributed in the hope that it will be
 *  useful, but WITHOUT ANY WARRANTY; without even the implied warranty
 *  of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with xrootd4j-cms-plugin.  If not, see
 *  <http://www.gnu.org/licenses/.>
 *
 */
package org.dcache.xrootd4j

import org.scalatest._
import junit.JUnitRunner
import org.junit.runner.RunWith

@RunWith(classOf[JUnitRunner])
class CmsMappingHandlerTest extends FlatSpec {

  val map = CmsSettingsParser.parse(
    <storage-mapping>
      <lfn-to-pfn protocol="direct" path-match="sub1/(.*)" result="subchain1/$1"/>
      <lfn-to-pfn protocol="direct" path-match="sub2/(.*)" result="subchain2/$1"/>
      <lfn-to-pfn protocol="direct" path-match="role/path/(.*)" result="correctpath/$1"/>
      <lfn-to-pfn protocol="root" path-match="/simple/(.*)" result="$1"/>
      <lfn-to-pfn protocol="root" chain="direct" path-match="/chained/(.*)" result="/chainbase/$1"/>
      <lfn-to-pfn protocol="root" path-match="/chained/role/(.*)" result="/wrongpath/$1"/>
      <lfn-to-pfn protocol="root" path-match="/one/(.*)/two/(.*)" result="$1$2"/>
    </storage-mapping>)

  val handler = new CmsMappingHandler(map, "root")
  val mappedFilenameFor = handler.authorize(null, null, null, _ : String, null, 0, null)

  it should "leave an unmatched path unchanged" in {
    expect("root://unmatched/blubb") {
      mappedFilenameFor("root://unmatched/blubb")
    }
  }

  it should "map a matching path" in {
    expect("wurstbrot") {
      mappedFilenameFor("/simple/wurstbrot")
    }
  }

  it should "map a matching chained rule" in {
    expect("/chainbase/subchain1/wurstbrot") {
      mappedFilenameFor("/chained/sub1/wurstbrot")
    }
  }

  it should "use the first matching rule" in {
    expect("/chainbase/correctpath/wurstbrot") {
      mappedFilenameFor("/chained/role/path/wurstbrot")
    }
  }

  it should "be able to understand patterns with two groups" in {
    expect("firstsecond") {
      mappedFilenameFor("/one/first/two/second")
    }
  }
}
